usr/opt/OSCX200/src/man/XI/XChgPtr.man
.\"
.\" *****************************************************************
.\" *                                                               *
.\" *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
.\" *                                                               *
.\" *   All Rights Reserved.  Unpublished rights  reserved  under   *
.\" *   the copyright laws of the United States.                    *
.\" *                                                               *
.\" *   The software contained on this media  is  proprietary  to   *
.\" *   and  embodies  the  confidential  technology  of  Digital   *
.\" *   Equipment Corporation.  Possession, use,  duplication  or   *
.\" *   dissemination of the software and media is authorized only  *
.\" *   pursuant to a valid written license from Digital Equipment  *
.\" *   Corporation.                                                *
.\" *                                                               *
.\" *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
.\" *   by the U.S. Government is subject to restrictions  as  set  *
.\" *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
.\" *   or  in  FAR 52.227-19, as applicable.                       *
.\" *                                                               *
.\" *****************************************************************
.\"
.\"
.\" HISTORY
.\"
.ds xL Programming With Xlib
.TH XChangePointerDevice 3X11 "Release 5" "X Version 11" "X FUNCTIONS"
.SH NAME
XChangePointerDevice \- change which device is the X pointer
.SH SYNTAX
Status XChangePointerDevice\^(\^\fIdisplay\fP, \fIdevice\fP\^ \fIxaxis\fP\^, \fIyaxis\fP\^)
.br
      Display *\fIdisplay\fP\^;
.br
      XDevice *\fIdevice\fP\^; 
.br
      int \fIxaxis\fP\^; 
.br
      int \fIyaxis\fP\^; 
.br
.SH ARGUMENTS
.TP 12
.I display
Specifies the connection to the X server.
.TP 12
.I device
Specifies the device to be used as the X pointer.
.TP 12
.I xaxis
Specifies the axis of the device to be used as the X pointer x-axis.
.TP 12
.I yaxis
Specifies the axis of the device to be used as the X pointer y-axis.
.SH DESCRIPTION
The \fIXChangePointerDevice\fP request causes the server to use the specified
device as the X pointer.  The device must have been opened by the client via
\fIXOpenDevice\fP or a \fIBadDevice\fP error will result.  The device must
support input class \fIValuators\fP or a \fIBadMatch\fP error will result.
If the implementation does not support use of the specified device as the
X pointer, a \fIBadDevice\fP error will result.
.P
If the specified device is grabbed by another client, \fIAlreadyGrabbed\fP
is returned.  If the specified device is frozen by a grab on another device,
\fIGrabFrozen\fP is returned.
If the request is successful, \fISuccess\fP is returned.
.P
If the request succeeds,
a \fIChangeDeviceNotify\fP event is sent to all clients that have selected that
event.  A \fIMappingNotify\fP event with request = \fIMappingPointer\fP is
sent to all clients.
The specified device becomes the X pointer, and
the old X pointer becomes accessible through the input extension
protocol requests.
.P
\fIXChangePointerDevice\fP can generate a \fIBadDevice\fP or a \fIBadMatch\fP
error.
.SH DIAGNOSTICS
.TP 12
\fIBadDevice\fP
An invalid device was specified.  The specified device does not exist, has 
not been opened by this client via \fIXOpenInputDevice\fP, or is already
one of the core X input devices (pointer or keyboard).  This error may also
occur if the server implementation does not support using the specified
device as the X pointer.
.TP 12
\fIBadMatch\fP
This error may occur if an \fIXChangePointerDevice\fP request was made 
specifying a device that has less than two valuators, or specifying a 
valuator index beyond the range supported by the device.
.SH "SEE ALSO"
XChangeKeyboardDevice
.br
\fI\*(xL\fP


usr/opt/OSCX200/src/extensions/server/xinput/xchgptr.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
/* $Header: /usr/sde/x11/rcs/x11/src/./extensions/server/xinput/xchgptr.c,v 1.2 91/12/15 12:42:16 devrcs Exp $ */

/************************************************************
Copyright (c) 1989 by Hewlett-Packard Company, Palo Alto, California, and the 
Massachusetts Institute of Technology, Cambridge, Massachusetts.

			All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the names of Hewlett-Packard or MIT not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.

HEWLETT-PACKARD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
HEWLETT-PACKARD BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

********************************************************/

/***********************************************************************
 *
 * Extension function to change the pointer device.
 *
 */

#define	 NEED_EVENTS
#define	 NEED_REPLIES
#include "X.h"				/* for inputstr.h    */
#include "Xproto.h"			/* Request macro     */
#include "XI.h"
#include "XIproto.h"
#include "inputstr.h"			/* DeviceIntPtr	     */
#include "windowstr.h"			/* window structure  */
#include "scrnintstr.h"			/* screen structure  */

extern	int 		IReqCode;
extern	int 		BadDevice;
extern	int 		ChangeDeviceNotify;
extern	Mask		ChangeDeviceNotifyMask;
extern	InputInfo	inputInfo;
extern	ScreenInfo	screenInfo;
extern	WindowPtr	*WindowTable;
extern	void		(* ReplySwapVector[256]) ();
DeviceIntPtr		LookupDeviceIntRec();

/***********************************************************************
 *
 * This procedure is invoked to swap the request bytes if the server and
 * client have a different byte order.
 *
 */

int
SProcXChangePointerDevice(client)
    register ClientPtr client;
    {
    register char n;

    REQUEST(xChangePointerDeviceReq);
    swaps(&stuff->length, n);
    return(ProcXChangePointerDevice(client));
    }

/***********************************************************************
 *
 * This procedure changes the device used as the X pointer.
 *
 */

ProcXChangePointerDevice (client)
    register ClientPtr client;
    {
    DeviceIntPtr 	xptr = inputInfo.pointer;
    DeviceIntPtr 	dev;
    ValuatorClassPtr 	v;
    xChangePointerDeviceReply	rep;
    changeDeviceNotify	ev;

    REQUEST(xChangePointerDeviceReq);
    REQUEST_SIZE_MATCH(xChangePointerDeviceReq);

    rep.repType = X_Reply;
    rep.RepType = X_ChangePointerDevice;
    rep.length = 0;
    rep.sequenceNumber = client->sequence;

    dev = LookupDeviceIntRec (stuff->deviceid);
    if (dev == NULL)
	{
	rep.status = -1;
	SendErrorToClient(client, IReqCode, X_ChangePointerDevice, 0, 
	    BadDevice);
	return Success;
	}

    v = dev->valuator;
    if (v == NULL || v->numAxes < 2 || 
	stuff->xaxis >= v->numAxes ||
	stuff->yaxis >= v->numAxes)
	{
	rep.status = -1;
	SendErrorToClient(client, IReqCode, X_ChangePointerDevice, 0, BadMatch);
	return Success;
	}

    if (((dev->grab) && !SameClient(dev->grab, client)) ||
        ((xptr->grab) && !SameClient(xptr->grab, client)))
	rep.status = AlreadyGrabbed;
    else if ((dev->sync.frozen &&
	     dev->sync.other && !SameClient(dev->sync.other, client)) ||
	     (xptr->sync.frozen &&
	      xptr->sync.other && !SameClient(xptr->sync.other, client)))
	rep.status = GrabFrozen;
    else
	{
	if (ChangePointerDevice (
	    xptr, dev, stuff->xaxis, stuff->yaxis) != Success)
	    {
	    SendErrorToClient(client, IReqCode, X_ChangePointerDevice, 0, 
		BadDevice);
	    return Success;
	    }
	if (dev->focus)
	    DeleteFocusClassDeviceStruct(dev);
	RegisterOtherDevice (xptr);
	RegisterPointerDevice (dev);

	ev.type = ChangeDeviceNotify;
	ev.deviceid = stuff->deviceid;
	ev.time = currentTime.milliseconds;
	ev.request = NewPointer;

	SendEventToAllWindows (dev, ChangeDeviceNotifyMask, &ev, 1);
	SendMappingNotify (MappingPointer, 0, 0);

	rep.status = 0;
	}

    WriteReplyToClient (client, sizeof (xChangePointerDeviceReply), 
	&rep);
    return Success;
    }

DeleteFocusClassDeviceStruct(dev)
    DeviceIntPtr dev;
    {
    xfree(dev->focus->trace);
    xfree(dev->focus);
    dev->focus = NULL;
    }

/***********************************************************************
 *
 * Send an event to interested clients in all windows on all screens.
 *
 */

SendEventToAllWindows (dev, mask, ev, count)
    DeviceIntPtr dev;
    Mask mask;
    xEvent *ev;
    int count;
    {
    int i;
    WindowPtr pWin, p1;

    for (i=0; i<screenInfo.numScreens; i++)
	{
	pWin = WindowTable[i];
	(void)DeliverEventsToWindow(pWin, ev, count, mask, NullGrab, dev->id);
	p1 = pWin->firstChild;
	FindInterestedChildren (dev, p1, mask, ev, count);
	}
    }

/***********************************************************************
 *
 * Walk through the window tree, finding all clients that want to know
 * about the ChangeDeviceNotify Event.
 *
 */

FindInterestedChildren (dev, p1, mask, ev, count)
    DeviceIntPtr	dev;
    WindowPtr 		p1;
    Mask		mask;
    xEvent		*ev;
    int			count;
    {
    WindowPtr p2;

    while (p1)
        {
        p2 = p1->firstChild;
	(void)DeliverEventsToWindow(p1, ev, count, mask, NullGrab, dev->id);
	FindInterestedChildren (dev, p2, mask, ev, count);
	p1 = p1->nextSib;
        }
    }

/***********************************************************************
 *
 * This procedure writes the reply for the XChangePointerDevice 
 * function, if the client and server have a different byte ordering.
 *
 */

SRepXChangePointerDevice (client, size, rep)
    ClientPtr	client;
    int		size;
    xChangePointerDeviceReply	*rep;
    {
    register char n;

    swaps(&rep->sequenceNumber, n);
    swapl(&rep->length, n);
    WriteToClient(client, size, rep);
    }
